include/dsn/cpp/json_helper.h

```diff
@@ -339,40 +339,21 @@ ENUM_TYPE_SERIALIZATION(dsn::replication::bulk_load_status::type,
                         dsn::replication::bulk_load_status::BLS_INVALID)
 
 // json serialization for gpid, we treat it as string: "app_id.partition_id"
-inline void json_encode(JsonWriter &out, const dsn::gpid &pid)
-{
-    json_encode(out, pid.to_string());
-}
-inline bool json_decode(const dsn::json::JsonObject &in, dsn::gpid &pid)
-{
-    std::string gpid_message;
-    dverify(json_decode(in, gpid_message));
-    return pid.parse_from(gpid_message.c_str());
-}
+void json_encode(JsonWriter &out, const dsn::gpid &pid);
+bool json_decode(const dsn::json::JsonObject &in, dsn::gpid &pid);
```

Comment:
Move the declaration to `gpid.h`.

```suggestion
namespace json {
class JsonWriter;
class JsonObject;
} // namespace json
void json_encode(JsonWriter &out, const gpid &pid);
bool json_decode(const JsonObject &in, gpid &pid);
```

Do the same to other types.

Reply:
Why move the declaration to `gpid.h`. I think placed them centrally in one file is ok.

Reply:
The module itself should be responsible for its related code. There's no modularity if everything is in a central file.